The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9693217257/job/26748208309) [(plain)](https://github.com/rust-lang/rust/commit/b959a23eb524f59733fafac7bed6d1005eb9ff77/checks/26748208309/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Getting action download info
Download action repository 'msys2/setup-msys2@v2.22.0' (SHA:cc11e9188b693c2b100158c3322424c4cc1dadea)
Download action repository 'actions/checkout@v4' (SHA:692973e3d937129bcbf40652eb9f2f61becf3332)
Download action repository 'actions/upload-artifact@v4' (SHA:65462800fd760344b1a7b4382951275a0abb4808)
Complete job name: PR - mingw-check-tidy
git config --global core.autocrlf false
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
---
COPY scripts/sccache.sh /scripts/
RUN sh /scripts/sccache.sh

COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt \
    && pip3 install virtualenv
COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---

#12 [5/8] COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
#12 DONE 0.0s

#13 [6/8] RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt     && pip3 install virtualenv
#13 0.429   Downloading binaryornot-0.4.4-py2.py3-none-any.whl (9.0 kB)
#13 0.445 Collecting boolean-py==4.0
#13 0.453   Downloading boolean.py-4.0-py3-none-any.whl (25 kB)
#13 0.470 Collecting chardet==5.1.0
---
#13 3.773 Building wheels for collected packages: reuse
#13 3.774   Building wheel for reuse (pyproject.toml): started
#13 4.092   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 4.093   Created wheel for reuse: filename=reuse-1.1.0-cp310-cp310-manylinux_2_35_x86_64.whl size=181117 sha256=f5f58750481f69515c2c0d1d503daf565e2565c370d07fc6aeb95fe3498b4269
#13 4.093   Stored in directory: /tmp/pip-ephem-wheel-cache-mlud4sdi/wheels/c2/3c/b9/1120c2ab4bd82694f7e6f0537dc5b9a085c13e2c69a8d0c76d
#13 4.096 Installing collected packages: boolean-py, binaryornot, setuptools, reuse, python-debian, markupsafe, license-expression, jinja2, chardet
#13 4.117   Attempting uninstall: setuptools
#13 4.118     Found existing installation: setuptools 59.6.0
#13 4.119     Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
---
#13 5.346   Downloading virtualenv-20.26.3-py3-none-any.whl (5.7 MB)
#13 5.586      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.7/5.7 MB 24.0 MB/s eta 0:00:00
#13 5.645 Collecting filelock<4,>=3.12.2
#13 5.652   Downloading filelock-3.15.4-py3-none-any.whl (16 kB)
#13 5.685 Collecting platformdirs<5,>=3.9.1
#13 5.693   Downloading platformdirs-4.2.2-py3-none-any.whl (18 kB)
#13 5.713 Collecting distlib<1,>=0.3.7
#13 5.720   Downloading distlib-0.3.8-py2.py3-none-any.whl (468 kB)
#13 5.731      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 468.9/468.9 KB 54.7 MB/s eta 0:00:00
#13 5.820 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 5.990 Successfully installed distlib-0.3.8 filelock-3.15.4 platformdirs-4.2.2 virtualenv-20.26.3
#13 DONE 6.1s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
DirectMap4k:      233408 kB
DirectMap2M:     8155136 kB
DirectMap1G:    10485760 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
    Finished `dev` profile [unoptimized] target(s) in 0.03s
##[endgroup]
downloading https://ci-artifacts.rust-lang.org/rustc-builds-alt/536235f07e57c9108c6c3b1eacb323164e0f4cfb/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/llvm-536235f07e57c9108c6c3b1eacb323164e0f4cfb-true/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm
---
   Compiling tidy v0.1.0 (/checkout/src/tools/tidy)
    Finished `release` profile [optimized] target(s) in 28.80s
##[endgroup]
fmt check
##[error]Diff in /checkout/compiler/rustc_mir_transform/src/jump_threading.rs at line 201:
         let mut state = State::new(ConditionSet::default(), self.map);
 
         let conds = if let Some((value, then, else_)) = targets.as_static_if() {
-            let value = ScalarInt::try_from_uint(value, discr_layout.size) ?;
+            let value = ScalarInt::try_from_uint(value, discr_layout.size)?;
             self.arena.alloc_from_iter([
                 Condition { value, polarity: Polarity::Eq, target: then },
                 Condition { value, polarity: Polarity::Ne, target: else_ },
fmt error: Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/rustfmt/bin/rustfmt" "--config-path" "/checkout" "--edition" "2021" "--unstable-features" "--skip-children" "--check" "/checkout/compiler/rustc_mir_transform/src/coverage/tests.rs" "/checkout/compiler/rustc_target/src/spec/base/illumos.rs" "/checkout/compiler/rustc_mir_transform/src/coverage/graph.rs" "/checkout/compiler/rustc_target/src/spec/base/avr_gnu.rs" "/checkout/compiler/rustc_target/src/spec/base/fuchsia.rs" "/checkout/compiler/rustc_mir_transform/src/coverage/mod.rs" "/checkout/compiler/rustc_target/src/spec/base/hurd_gnu.rs" "/checkout/compiler/rustc_mir_transform/src/coverage/query.rs" "/checkout/compiler/rustc_target/src/spec/base/bpf.rs" "/checkout/compiler/rustc_target/src/spec/base/wasm.rs" "/checkout/compiler/rustc_target/src/spec/base/unikraft_linux_musl.rs" "/checkout/compiler/rustc_target/src/spec/base/l4re.rs" "/checkout/compiler/rustc_target/src/spec/base/nto_qnx.rs" "/checkout/compiler/rustc_target/src/spec/base/thumb.rs" "/checkout/compiler/rustc_target/src/spec/base/linux_gnu.rs" "/checkout/compiler/rustc_target/src/spec/base/openbsd.rs" "/checkout/compiler/rustc_target/src/spec/base/linux_ohos.rs" "/checkout/compiler/rustc_target/src/spec/base/teeos.rs" "/checkout/compiler/rustc_mir_transform/src/coverage/spans/from_mir.rs" "/checkout/compiler/rustc_mir_transform/src/coverage/mappings.rs" "/checkout/compiler/rustc_mir_transform/src/coverage/counters.rs" "/checkout/compiler/rustc_mir_transform/src/coverage/spans.rs" "/checkout/compiler/rustc_mir_transform/src/reveal_all.rs" "/checkout/compiler/rustc_mir_transform/src/cost_checker.rs" "/checkout/compiler/rustc_mir_transform/src/function_item_references.rs" "/checkout/compiler/rustc_mir_transform/src/jump_threading.rs" "/checkout/compiler/rustc_target/src/spec/abi/tests.rs" "/checkout/compiler/rustc_mir_transform/src/required_consts.rs" "/checkout/compiler/rustc_target/src/spec/abi/mod.rs" "/checkout/compiler/rustc_mir_transform/src/shim/async_destructor_ctor.rs" "/checkout/compiler/rustc_mir_transform/src/shim.rs" "/checkout/compiler/rustc_mir_transform/src/errors.rs" "/checkout/compiler/rustc_mir_transform/src/elaborate_box_derefs.rs" "/checkout/compiler/rustc_mir_transform/src/ctfe_limit.rs" "/checkout/compiler/rustc_mir_transform/src/remove_noop_landing_pads.rs" "/checkout/compiler/rustc_mir_transform/src/gvn.rs" "/checkout/compiler/rustc_mir_transform/src/add_subtyping_projections.rs" "/checkout/compiler/rustc_mir_transform/src/elaborate_drops.rs" "/checkout/compiler/rustc_graphviz/src/tests.rs" "/checkout/compiler/rustc_graphviz/src/lib.rs" "/checkout/compiler/rustc_middle/src/tests.rs" "/checkout/compiler/rustc_middle/src/thir.rs" "/checkout/compiler/rustc_middle/src/macros.rs" "/checkout/compiler/rustc_middle/src/dep_graph/mod.rs" "/checkout/compiler/rustc_middle/src/dep_graph/dep_node.rs" "/checkout/compiler/rustc_middle/src/middle/stability.rs" "/checkout/compiler/rustc_middle/src/middle/privacy.rs" "/checkout/compiler/rustc_middle/src/middle/mod.rs" "/checkout/compiler/rustc_middle/src/middle/dependency_format.rs" "/checkout/compiler/rustc_middle/src/middle/lang_items.rs" "/checkout/compiler/rustc_middle/src/middle/region.rs" "/checkout/compiler/rustc_middle/src/middle/exported_symbols.rs" "/checkout/compiler/rustc_middle/src/middle/limits.rs" "/checkout/compiler/rustc_middle/src/middle/resolve_bound_vars.rs" "/checkout/compiler/rustc_middle/src/middle/codegen_fn_attrs.rs" "/checkout/compiler/rustc_middle/src/middle/debugger_visualizer.rs" "/checkout/compiler/rustc_middle/src/lint.rs" "/checkout/compiler/rustc_middle/src/util/call_kind.rs" "/checkout/compiler/rustc_middle/src/util/mod.rs" "/checkout/compiler/rustc_middle/src/util/common.rs" "/checkout/compiler/rustc_middle/src/util/find_self_call.rs" "/checkout/compiler/rustc_middle/src/util/common/tests.rs" "/checkout/compiler/rustc_middle/src/util/bug.rs" "/checkout/compiler/rustc_target/src/spec/base/linux.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
  local time: Thu Jun 27 08:32:07 UTC 2024
  network time: Thu, 27 Jun 2024 08:32:07 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

--- a/compiler/rustc_mir_dataflow/src/value_analysis.rs
+++ b/compiler/rustc_mir_dataflow/src/value_analysis.rs
@@ -846,9 +846,10 @@ impl Map {
 
         if let ty::Ref(_, ref_ty, _) | ty::RawPtr(ref_ty, _) = ty.kind()
             && let ty::Slice(..) = ref_ty.kind()
+            // The user may have written a predicate like `[T]: Sized` in their where clauses,
+            // which makes slices scalars.
+            && self.places[place].value_index.is_none()
         {
-            assert!(self.places[place].value_index.is_none(), "slices are not scalars");
-
             // Prepend new child to the linked list.
             let len = self.places.push(PlaceInfo::new(Some(TrackElem::DerefLen)));
             self.places[len].next_sibling = self.places[place].first_child;

--- a/compiler/rustc_mir_transform/src/jump_threading.rs
+++ b/compiler/rustc_mir_transform/src/jump_threading.rs
@@ -270,12 +270,13 @@ impl<'tcx, 'a> TOFinder<'tcx, 'a> {
                     self.process_switch_int(discr, targets, bb, &mut state);
                     self.find_opportunity(pred, state, cost, depth + 1);
                 }
-                _ => self.recurse_through_terminator(pred, &state, &cost, depth),
+                _ => self.recurse_through_terminator(pred, || state, &cost, depth),
             }
-        } else {
+        } else if let &[ref predecessors @ .., last_pred] = &predecessors[..] {
             for &pred in predecessors {
-                self.recurse_through_terminator(pred, &state, &cost, depth);
+                self.recurse_through_terminator(pred, || state.clone(), &cost, depth);
             }
+            self.recurse_through_terminator(last_pred, || state, &cost, depth);
         }
 
         let new_tos = &mut self.opportunities[last_non_rec..];
@@ -566,11 +567,12 @@ impl<'tcx, 'a> TOFinder<'tcx, 'a> {
         None
     }
 
-    #[instrument(level = "trace", skip(self, cost))]
+    #[instrument(level = "trace", skip(self, state, cost))]
     fn recurse_through_terminator(
         &mut self,
         bb: BasicBlock,
-        state: &State<ConditionSet<'a>>,
+        // Pass a closure that may clone the state, as we don't want to do it each time.
+        state: impl FnOnce() -> State<ConditionSet<'a>>,
         cost: &CostChecker<'_, 'tcx>,
         depth: usize,
     ) {
@@ -600,7 +602,7 @@ impl<'tcx, 'a> TOFinder<'tcx, 'a> {
         };
 
         // We can recurse through this terminator.
-        let mut state = state.clone();
+        let mut state = state();
         if let Some(place_to_flood) = place_to_flood {
             state.flood_with(place_to_flood.as_ref(), self.map, ConditionSet::default());
         }

--- a/tests/ui/mir/sized-slice-predicate-116721.rs
+++ b/tests/ui/mir/sized-slice-predicate-116721.rs
@@ -1,5 +1,6 @@
-//@ known-bug: #116721
+//@ build-pass
 //@ compile-flags: -Zmir-opt-level=3 --emit=mir
+
 fn hey<T>(it: &[T])
 where
     [T]: Clone,

